File: src/main/kotlin/xyz/mahmoudahmed/dsl/TypedBranchConditionBuilder.java
package xyz.mahmoudahmed.dsl;

import java.util.function.Predicate;

public class TypedBranchConditionBuilder<S, T> {
    private final TypedBranchBuilder<S, T> parent;
    private final Predicate<S> condition;

    TypedBranchConditionBuilder(TypedBranchBuilder<S, T> parent, Predicate<S> condition) {
        this.parent = parent;
        this.condition = condition;
    }

    public TypedBranchActionBuilder<S, T> then() {
        return new TypedBranchActionBuilder<>(parent, (Predicate<Object>)(Predicate<?>) condition);
    }
}

File: src/main/kotlin/xyz/mahmoudahmed/dsl/typed/TypedBranchConditionBuilder.kt
package xyz.mahmoudahmed.dsl.typed

class TypedBranchConditionBuilder<S, T>(
    private val parent: TypedBranchBuilder<S, T>,
    private val condition: (S) -> Boolean
) {
    fun then(): TypedBranchActionBuilder<S, T> {
        return TypedBranchActionBuilder(parent) { obj ->
            @Suppress("UNCHECKED_CAST")
            condition(obj as S)
        }
    }
}